prow/hook/server.go

```diff
@@ -17,238 +17,433 @@ limitations under the License.
 package hook
 
 import (
-	"bytes"
-	"encoding/json"
-	"fmt"
-	"io/ioutil"
-	"net/http"
-	"strconv"
-	"strings"
 	"sync"
 	"time"
 
+	"github.com/prometheus/client_golang/prometheus"
 	"github.com/sirupsen/logrus"
 
 	"k8s.io/test-infra/prow/config"
 	"k8s.io/test-infra/prow/github"
 	_ "k8s.io/test-infra/prow/hook/plugin-imports"
 	"k8s.io/test-infra/prow/plugins"
+	"k8s.io/test-infra/prow/servemux"
+)
+
+const (
+	failedCommentCoerceFmt = "Could not coerce %s event to a GenericCommentEvent. Unknown 'action': %q."
+	eventTypeField         = "event-type"
+)
+
+var (
+	nonCommentIssueActions = map[github.IssueEventAction]bool{
+		github.IssueActionAssigned:     true,
+		github.IssueActionUnassigned:   true,
+		github.IssueActionLabeled:      true,
+		github.IssueActionUnlabeled:    true,
+		github.IssueActionMilestoned:   true,
+		github.IssueActionDemilestoned: true,
+		github.IssueActionClosed:       true,
+		github.IssueActionReopened:     true,
+		github.IssueActionPinned:       true,
+		github.IssueActionUnpinned:     true,
+		github.IssueActionTransferred:  true,
+		github.IssueActionDeleted:      true,
+		github.IssueActionLocked:       true,
+		github.IssueActionUnlocked:     true,
+	}
+	nonCommentPullRequestActions = map[github.PullRequestEventAction]bool{
+		github.PullRequestActionAssigned:             true,
+		github.PullRequestActionUnassigned:           true,
+		github.PullRequestActionReviewRequested:      true,
+		github.PullRequestActionReviewRequestRemoved: true,
+		github.PullRequestActionLabeled:              true,
+		github.PullRequestActionUnlabeled:            true,
+		github.PullRequestActionClosed:               true,
+		github.PullRequestActionReopened:             true,
+		github.PullRequestActionSynchronize:          true,
+		github.PullRequestActionReadyForReview:       true,
+		github.PullRequestConvertedToDraft:           true,
+	}
 )
 
 // Server implements http.Handler. It validates incoming GitHub webhooks and
 // then dispatches them to the appropriate plugins.
 type Server struct {
-	ClientAgent    *plugins.ClientAgent
-	Plugins        *plugins.ConfigAgent
-	ConfigAgent    *config.Agent
-	TokenGenerator func() []byte
-	Metrics        *Metrics
-
-	// c is an http client used for dispatching events
-	// to external plugin services.
-	c http.Client
-	// Tracks running handlers for graceful shutdown
-	wg sync.WaitGroup
+	ClientAgent *plugins.ClientAgent
```

Comment:
Why does hook not use the `servemux` server and has its own instead?

Reply:
oops. that was a leftover of something i was trying to check. I will remove it. Thanks.

Reply:
I deleted the previous comment since I was referring to a different thing.  So for this one, the `servemux` server shouldn't care about having this.